### PR TITLE
Add yaml for data.gov catalog

### PR DESCRIPTION
**Pull request by avdata99**
Using OKF data.json extension and removing USMetadata extension

### CODE REVIEW

File: .docker-compose.datagov-catalog.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
+version: '3.2'
```
Comment (akariv) [inline .docker-compose.datagov-catalog.yaml]: Please modify the  `.docker-compose.datagov.yaml` for modifying the base image and don't create a new base image flavour..

File: docker-compose.yaml [hunk 2]
```diff
@@ -39,6 +39,8 @@ services:
     restart: always
     expose:
     - "8080"
+    ports:
```
Comment (akariv) [inline docker-compose.yaml]: These ports are closed in the base file for security reasons.
Any port opening should be done in the flavour-specific (e.g. `datagov`) files.

File: docker-compose/ckan-conf-templates/datagov-catalog-production.ini.template [hunk 3]
```diff
@@ -0,0 +1,231 @@
+[DEFAULT]
```
Comment (akariv) [inline docker-compose/ckan-conf-templates/datagov-catalog-production.ini.template]: please modify the main datagov flavour production.ini template and don't create a new flavour.

**Review by akariv — CHANGES_REQUESTED** [review]
Left some comments.

Also note that the build failed, so better check that before resubmitting the PR.

File: .docker-compose.datagov-catalog.yaml (hunk 1 above)
Comment (avdata99) [reply .docker-compose.datagov-catalog.yaml]: I will use the previous yaml file but I'm thinking that we already need two flavours: Inventory and Catalog

File: .docker-compose.datagov-theme.yaml [hunk 4]
```diff
@@ -25,9 +27,8 @@ services:
           install_standard_ckan_extension_github -r ckan/ckanext-harvest &&\
           install_standard_ckan_extension_github -r datopian/ckanext-logstash &&\
           install_standard_ckan_extension_github -r ckan/ckanext-dcat ckanext-dcat &&\
-          install_standard_ckan_extension_github -r GSA/ckanext-geodatagov &&\
-          install_standard_ckan_extension_github -r akariv/ckanext-datajson -b datagov &&\
-          install_standard_ckan_extension_github -r akariv/USMetadata -e ckanext-usmetadata &&\
+          install_standard_ckan_extension_github -r datopian/ckanext-geodatagov &&\
+          install_standard_ckan_extension_github -r avdata99/ckanext-datajson -b datagov &&\
```
Comment (akariv) [inline .docker-compose.datagov-theme.yaml]: Aren't we using the okfn one in the end?

File: .docker-compose.datagov-theme.yaml [hunk 5]
```diff
@@ -25,9 +27,8 @@ services:
           install_standard_ckan_extension_github -r ckan/ckanext-harvest &&\
           install_standard_ckan_extension_github -r datopian/ckanext-logstash &&\
           install_standard_ckan_extension_github -r ckan/ckanext-dcat ckanext-dcat &&\
-          install_standard_ckan_extension_github -r GSA/ckanext-geodatagov &&\
-          install_standard_ckan_extension_github -r akariv/ckanext-datajson -b datagov &&\
-          install_standard_ckan_extension_github -r akariv/USMetadata -e ckanext-usmetadata &&\
```
Comment (akariv) [inline .docker-compose.datagov-theme.yaml:30]: We still need the USMetadata extension in non-harvester datagov instances.

File: create_secrets.py [hunk 6]
```diff
@@ -13,10 +13,13 @@
...
 def set_databse_urls(secrets):
-    write_secrets['ckan'].append(
-        'export SQLALCHEMY_URL=postgresql://ckan:{db_password}@db/ckan'.format(
+    alchemy_url = 'SQLALCHEMY_URL=postgresql://ckan:{db_password}@db/ckan'.format(
             db_password=secrets['db-POSTGRES_PASSWORD']
-    ))
+            )
+    write_secrets['ckan'].append('export {}'.format(alchemy_url))
+    # add as env file also
+    write_secrets['ckanenv'].append(alchemy_url)
```
Comment (akariv) [inline create_secrets.py]: What is this new file?
Why not name it with a non-confusing name (i.e. 'harvester')?

File: harvester/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,5 @@
+FROM viderum/ckan-ng-harvest:latest
```
Comment (akariv) [inline harvester/Dockerfile]: Why do we need this Dockerfile?

File: nginx/default.conf [hunk 8]
```diff
@@ -4,6 +4,28 @@ server {
...
     client_max_body_size 1000M;
+
+    location /airflow/ {
+        proxy_set_header Host $host;
+        proxy_redirect off;
+        proxy_http_version 1.1;
+        proxy_set_header Upgrade $http_upgrade;
+        proxy_set_header Connection "upgrade";
+
+        # If the proxy_pass statement has no variables in it, then it will use the "gethostbyaddr" system call during start-up or reload and will cache that value permanently.
```
Comment (akariv) [inline nginx/default.conf]: Best to remove these comments from production code.

File: .docker-compose.datagov-theme.yaml (hunk 4 above)
Comment (avdata99) [reply .docker-compose.datagov-theme.yaml]: No, OFK data.json extension is not updated. [I updated merging different forks](https://github.com/avdata99/ckanext-datajson/blob/datagov/cherry-picking-from-other-forks.md)
GeoDataGov was forked in order to test it and the send PR to GSA

File: .docker-compose.datagov-theme.yaml (hunk 5 above)
Comment (avdata99) [reply .docker-compose.datagov-theme.yaml:30]: I know, I proposed to create a new yaml file for the **inventory**.
This yaml file is intented to work as **catalog**

File: create_secrets.py (hunk 6 above)
Comment (avdata99) [reply create_secrets.py]: This file include the CKAN secrets as _env_ file (without the _EXPORT_). Use the name you prefer. This file could be used for any container, no just harvester.

File: harvester/Dockerfile (hunk 7 above)
Comment (avdata99) [reply harvester/Dockerfile]: My first try to get the CKAN API KEY was via Dockerfile. I'll delete this file

File: nginx/default.conf (hunk 8 above)
Comment (avdata99) [reply nginx/default.conf]: removed

File: .docker-compose.datagov-theme.yaml (hunk 4 above)
Comment (akariv) [reply .docker-compose.datagov-theme.yaml]: So can we move our datajson extension to the datopian organisation?

File: .docker-compose.datagov-theme.yaml (hunk 5 above)
Comment (akariv) [reply .docker-compose.datagov-theme.yaml:30]: What I mean is that we need this extension _installed_ - we won't _enable_ it in the harvester kind.

File: create_secrets.py (hunk 6 above)
Comment (akariv) [reply create_secrets.py]: If it's only used by harvester, then name it accordingly.
e.g. the ckan secrets are named `ckan` as they're being used for ckan

File: docs/notes-about-airflow-nginx.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
+# Notes about Airflow on nginx
```
Comment (akariv) [inline docs/notes-about-airflow-nginx.md]: These are from the nginx help file.. .why do we need this here?

**Comment by akariv** [comment]
Note that the build is still failing, this time I think it's related to your modifications
```
ERROR: Couldn't find env file: /home/travis/build/ViderumGlobal/ckan-cloud-docker/docker-compose/ckanenv-secrets.sh
```

**Comment by aluminiumgeek** [comment]
> Note that the build is still failing, this time I think it's related to your modifications
> 
> ```
> ERROR: Couldn't find env file: /home/travis/build/ViderumGlobal/ckan-cloud-docker/docker-compose/ckanenv-secrets.sh
> ```

@avdata99 if `ckanenv-secrets.sh` is a new file (if it's not supposed to be `ckan-secrets.sh`), you need to append ckanenv secrets to `docker-compose/ckan-secrets.dat`. See the file for format reference:
each line should be like
`ckanenv <optional|required> ENV_VAR_NAME default_value Description`.

This file is used to generate secrets locally with `create_secrets.py` script.

File: .docker-compose.datagov-theme.yaml (hunk 4 above)
Comment (avdata99) [reply .docker-compose.datagov-theme.yaml]: We need to define who will be the maintainer of this ext.
If GSA will maintain this we can delete and OR to GSA repo

File: create_secrets.py (hunk 6 above)
Comment (avdata99) [reply create_secrets.py]: Changed for _harvester_

File: .docker-compose.datagov-theme.yaml (hunk 5 above)
Comment (avdata99) [reply .docker-compose.datagov-theme.yaml:30]: I added this ext again.

File: docs/notes-about-airflow-nginx.md (hunk 9 above)
Comment (avdata99) [reply docs/notes-about-airflow-nginx.md]: I moved to _nginx_ folder. Is this a good place for this?

File: .docker-compose.datagov-theme.yaml (hunk 4 above)
Comment (akariv) [reply .docker-compose.datagov-theme.yaml]: This is an official OKF extension, unlikely that GSA will be the maintainers of it.

File: docs/notes-about-airflow-nginx.md (hunk 9 above)
Comment (akariv) [reply docs/notes-about-airflow-nginx.md]: Can you summarize what this file is used for?

Comment (avdata99) [reply docs/notes-about-airflow-nginx.md]: https://github.com/avdata99/ckan-cloud-docker/blob/master/nginx/notes-about-airflow-nginx.md
in the _nginx_ folder

**Comment by avdata99** [comment]
> > ```
> > ERROR: Couldn't find env file: /home/travis/build/ViderumGlobal/ckan-cloud-docker/docker-compose/ckanenv-secrets.sh
> > ```
> 
> @avdata99 if `ckanenv-secrets.sh` is a new file (if it's not supposed to be `ckan-secrets.sh`), you need to append ckanenv secrets to `docker-compose/ckan-secrets.dat`. 

We don't need this reference at _ckan-secrets.dat_ since we just reuse the previous config values

**Comment by avdata99** [comment]
The error continues, now underestand why
```
ERROR: Couldn't find env file: /home/travis/build/ViderumGlobal/ckan-cloud-docker/docker-compose/harvester-secrets.sh
The command "bash .travis.sh script" exited with 1.
```
create_secrets script create the file but you are using the operator.
I'll try to add the harvester file in this operator scripts

**Comment by avdata99** [comment]
Look at this @akariv @aluminiumgeek 
https://github.com/avdata99/ckan-cloud-docker/commit/d594739a7d8df7bc85b19aff5d754d75563e7199
Maybe is the solution
I'm not sure where to run the `cca-operator initialize-harvester-secrets ckan-env-vars harvester-secrets` needed

File: docs/notes-about-airflow-nginx.md (hunk 9 above)
Comment (akariv) [reply docs/notes-about-airflow-nginx.md]: Not what's in the file (that I can see) - what is this file used for? Why is it here?

Comment (avdata99) [reply docs/notes-about-airflow-nginx.md]: I deleted the file.

**Comment by akariv** [comment]
@avdata99 that last commit is totally unrelated.

The build fails because this file is not created `/home/travis/build/ViderumGlobal/ckan-cloud-docker/docker-compose/harvester-secrets.sh`. 
What part of your code is supposed to create it?

File: .docker-compose.harvester_ng.yaml [hunk 10]
```diff
@@ -0,0 +1,64 @@
+version: '3.2'
```
Comment (akariv) [inline .docker-compose.harvester_ng.yaml]: This filename is not named correctly

It should be `.docker-compose-harvester_ng.yaml`
and not `.docker-compose.harvester_ng.yaml`

Comment (avdata99) [reply .docker-compose.harvester_ng.yaml]: image

What's the rule?

Comment (avdata99) [reply .docker-compose.harvester_ng.yaml]: [Changed](https://github.com/avdata99/ckan-cloud-docker/commit/171c4812e19aef25e6c4c02ea917220d24cae334)

**Comment by akariv** [comment]
.docker-compose.FLAVOUR.yaml is the naming for a flavour of our images,
usually targeted for a specific client The harvest_ng docker-compose file
is not a flavour, but an overlay of the docker compose used for local
development.

File: .docker-compose.datagov-theme.yaml (hunk 4 above)
Comment (akariv) [reply .docker-compose.datagov-theme.yaml]: @avdata99 any progress on moving this to the datopian organisation?

File: cca-operator/README.md [hunk 11]
```diff
@@ -103,6 +103,7 @@ Run the cca-operator CKAN commands:
 * create the ckan env vars secret: `cca-operator initialize-ckan-env-vars ckan-env-vars`
   * If you use the centralized infra, set the env vars: `--env CKAN_CLOUD_INSTANCE_ID=$CKAN_NAMESPACE --env CKAN_CLOUD_POSTGRES_HOST=db.ckan-cloud --env CKAN_CLOUD_POSTGRES_USER=postgres --env PGPASSWORD=123456 --env CKAN_CLOUD_SOLR_HOST=solr.ckan-cloud --env CKAN_CLOUD_SOLR_PORT=8983`
 * Initialize the CKAN secrets.sh: `cca-operator initialize-ckan-secrets ckan-env-vars ckan-secrets`
+* Initialize the Harvester secrets.sh: `cca-operator initialize-harvester-secrets ckan-env-vars harvester-secrets`
```
Comment (akariv) [inline cca-operator/README.md]: Not needed

File: cca-operator/cca-operator.sh [hunk 12]
```diff
@@ -74,14 +74,42 @@ if [ "${1}" == "initialize-ckan-env-vars" ]; then
         echo Ckan env vars secret already exists && exit 0
     fi
 
+elif [ "${1}" == "initialize-harvester-secrets" ]; then
```
Comment (akariv) [inline cca-operator/cca-operator.sh]: Not needed

File: create_secrets.py [hunk 13]
```diff
@@ -13,10 +13,16 @@
...
-    write_secrets['ckan'].append(
-        'export SQLALCHEMY_URL=postgresql://ckan:{db_password}@db/ckan'.format(
+    alchemy_url = 'SQLALCHEMY_URL=postgresql://ckan:{db_password}@db/ckan'.format(
             db_password=secrets['db-POSTGRES_PASSWORD']
-    ))
+            )
+    write_secrets['ckan'].append('export {}'.format(alchemy_url))
+    
+    # add as env file also
+    if 'harvester' not in write_secrets:
```
Comment (akariv) [inline create_secrets.py]: use 'setdefault' here instead.

File: start-harvester.md [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+### Other tools
+
+Clean all the data at the images
+
+```
+docker-compose \
+      -f docker-compose.yaml \
+      -f .docker-compose-db.yaml \
+      -f .docker-compose.datagov-theme.yaml \
+      -f .docker-compose.harvester_ng.yaml \
```
Comment (akariv) [inline start-harvester.md]: fix filename

Comment (akariv) [reply start-harvester.md]: (multiple locations)

**Review by akariv — CHANGES_REQUESTED** [review]
@avdata99 a few final fixes (I hope)

**Comment by avdata99** [comment]
> @avdata99 that last commit is totally unrelated.
> 
> The build fails because this file is not created `/home/travis/build/ViderumGlobal/ckan-cloud-docker/docker-compose/harvester-secrets.sh`.
> What part of your code is supposed to create it?

I also tried this:
https://github.com/avdata99/ckan-cloud-docker/commit/ffa97e1c46f2e023478d6dfa033b5a5919f3bb31

and this: 
https://github.com/avdata99/ckan-cloud-docker/commit/e4d544ededf93f4511223cf6605febab72a6f24d

So I think _create_secrets_ is only for run locally. 
How the secrets files are created when testing?

I didn't find it @akariv.

**Comment by akariv** [comment]
Andres didn't you see that the last file rename already fixed the ci?

**Comment by avdata99** [comment]
> Andres didn't you see that the last file rename already fixed the ci?

I just check, saw this new lines at _travis.sh_ and [implement them at my fork](https://github.com/avdata99/ckan-cloud-docker/commit/c34f45b57f3185ded630ac45ae1651ebda923466).

It worked.

**Comment by akariv** [comment]
@avdata99 - 
Never copy code from another fork like this.
Always use either `git pull` or (even better) `git rebase`.

File: .docker-compose.datagov-theme.yaml (hunk 4 above)
Comment (avdata99) [reply .docker-compose.datagov-theme.yaml]: @akariv 
Moved as Datopian repo: https://github.com/datopian/ckanext-datajson
Also migrated to CircleCI: https://circleci.com/gh/datopian/ckanext-datajson

File: cca-operator/cca-operator.sh (hunk 12 above)
Comment (avdata99) [reply cca-operator/cca-operator.sh]: Deleted

File: create_secrets.py (hunk 13 above)
Comment (avdata99) [reply create_secrets.py]: [done](https://github.com/avdata99/ckan-cloud-docker/commit/1d73a5a6f2b77f7fdd0148d7a31d538f1d80718f)

File: .docker-compose.datagov-theme.yaml [hunk 15]
```diff
@@ -25,8 +27,8 @@ services:
           install_standard_ckan_extension_github -r ckan/ckanext-harvest &&\
           install_standard_ckan_extension_github -r datopian/ckanext-logstash &&\
           install_standard_ckan_extension_github -r ckan/ckanext-dcat ckanext-dcat &&\
-          install_standard_ckan_extension_github -r GSA/ckanext-geodatagov &&\
-          install_standard_ckan_extension_github -r akariv/ckanext-datajson -b datagov &&\
+          install_standard_ckan_extension_github -r datopian/ckanext-geodatagov &&\
+          install_standard_ckan_extension_github -r avdata99/ckanext-datajson -b datagov &&\
```
Comment (adborden) [inline .docker-compose.datagov-theme.yaml]: Shouldn't these be the GSA repos?

Comment (avdata99) [reply .docker-compose.datagov-theme.yaml]: > Shouldn't these be the GSA repos?

I think they can be repos at GSA. I already sent a PR for GeoDataGov: https://github.com/GSA/ckanext-geodatagov/pull/120

**Comment by akariv** [comment]
@avdata99 

See last comments on setdefault usage
Also, let's move ckanext-datajson to datopian as well.

File: cca-operator/README.md (hunk 11 above)
Comment (avdata99) [reply cca-operator/README.md]: It was deleted

**Comment by avdata99** [comment]
Rebased and fixes write_secrets dict @akariv. Maybe we are ready to go
